src/vs/workbench/contrib/notebook/test/browser/contrib/notebookOutlineViewProviders.test.ts

```diff
@@ -0,0 +1,674 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as assert from 'assert';
+import { IDataSource } from 'vs/base/browser/ui/tree/tree';
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { mock } from 'vs/base/test/common/mock';
+import { ensureNoDisposablesAreLeakedInTestSuite } from 'vs/base/test/common/utils';
+import { ITextModel } from 'vs/editor/common/model';
+import { IOutlineModelService, OutlineModel } from 'vs/editor/contrib/documentSymbols/browser/outlineModel';
+import { TestConfigurationService } from 'vs/platform/configuration/test/common/testConfigurationService';
+import { TestThemeService } from 'vs/platform/theme/test/common/testThemeService';
+import { NotebookBreadcrumbsProvider, NotebookCellOutline, NotebookOutlinePaneProvider, NotebookQuickPickProvider } from 'vs/workbench/contrib/notebook/browser/contrib/outline/notebookOutline';
+import { ICellViewModel } from 'vs/workbench/contrib/notebook/browser/notebookBrowser';
+import { NotebookOutlineEntryFactory } from 'vs/workbench/contrib/notebook/browser/viewModel/notebookOutlineEntryFactory';
+import { OutlineEntry } from 'vs/workbench/contrib/notebook/browser/viewModel/OutlineEntry';
+import { INotebookExecutionStateService } from 'vs/workbench/contrib/notebook/common/notebookExecutionStateService';
+import { MockDocumentSymbol } from 'vs/workbench/contrib/notebook/test/browser/testNotebookEditor';
+import { OutlineTarget } from 'vs/workbench/services/outline/browser/outline';
+
+suite('Notebook Outline View Providers', function () {
+	// #region Setup
+	ensureNoDisposablesAreLeakedInTestSuite();
+
+	const configurationService = new TestConfigurationService();
+	const themeService = new TestThemeService();
+
+	const symbolsPerTextModel: Record<string, MockDocumentSymbol[]> = {};
+	function setSymbolsForTextModel(symbols: MockDocumentSymbol[], textmodelId = 'textId') {
+		symbolsPerTextModel[textmodelId] = symbols;
+	}
+
+	const executionService = new class extends mock<INotebookExecutionStateService>() {
+		override getCellExecution() { return undefined; }
+	};
+
+	class OutlineModelStub {
+		constructor(private textId: string) { }
+
+		getTopLevelSymbols() {
+			return symbolsPerTextModel[this.textId];
+		}
+	}
+	const outlineModelService = new class extends mock<IOutlineModelService>() {
+		override getOrCreate(model: ITextModel, arg1: any) {
+			const outline = new OutlineModelStub(model.id) as unknown as OutlineModel;
+			return Promise.resolve(outline);
+		}
+		override getDebounceValue(arg0: any) {
+			return 0;
+		}
+	};
+	// #endregion
+	// #region Helpers
+	function createCodeCellViewModel(version: number = 1, source = '# code', textmodelId = 'textId') {
+		return {
+			textBuffer: {
+				getLineCount() { return 0; }
```

Comment:
maybe we can make `source: string[]` so `getLineCount` can be accurate `source.length`